compiler/rustc_expand/src/expand.rs

```diff
@@ -1266,6 +1312,8 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
         }
 
         // The placeholder expander gives ids to statements, so we avoid folding the id here.
+        // We don't use `assign_id!` - it will be called when we visit statement's contents
+        // (e.g. an expression, item, or local)
```

Comment:
It may be more consistent to use `assign_id!` with a `Stmt` instead, because `Stmt` is an attribute target (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210).

Reply:
For `StmtKind::Item` and `StmtKind::Expr`, we will immediately overwrite the tracked statement ID with the item/expr id. Also, `Stmt` id assignment is performed in `PlaceholderExpander` instead of `InvocationCollector`, so using the current approach allows us to keep that logic as-is.